### DISCUSSION TITLE
How to prevent TextBox vertical scrollbar from covering text?

### DISCUSSION THREAD

**Discussion by TangDave** [AvaloniaUI/Avalonia#16995] [Q&A]
(Avalonia 11.1.2 on Windows) I have a TextBox with wrapped text long enough that a vertical scrollbar appears. Even when the scrollbar is 'minimized' (before being 'activated' by a mouse, for instance) it covers part of the TextBox content (part of a letter on each line) and when the scrollbar is activated even more is covered by the scrollbar (a few letters on each line). Is there a way to prevent this? I tried padding, but of course this doesn't work since the scrollbar is also padded, resulting in the same annoying coverage of text by the scrollbar.

**Comment by Wasdalt** [comment]
Use padding or hide the scrollbar

**Reply by TangDave** [reply]
Thanks for your response, but you really should just take a look at a textbox with a vertical scrollbar to see what I mean.

Padding also shifts the scrollbar to the left and covers the same amount of text. The scrollbar auto-hides, but that's not the point - when it does come into view it obscures text - many people can read fast enough that they are scrolling while they are reading - how do you do that when the scrollbar is hiding the text on the right side? Even when the scrollbar is 'hidden', a thin slice obscures part of the letter on the far right - not as critical but still annoying.

To reproduce, use a TextBox with wrapping (TextWrapping="Wrap") - put enough text in it to cause the vertical scrollbar to come into play. At runtime try to scroll slowly while reading the text.

**Reply by Wasdalt** [reply]
The option will work for itself, but if you don't dig into the Styles constructors. 

```
    <TextBox
        PointerWheelChanged="TextBox_PointerWheelChanged"
        ScrollViewer.AllowAutoHide="False"
        ScrollViewer.HorizontalScrollBarVisibility="Disabled"
        ScrollViewer.VerticalScrollBarVisibility="Hidden"
        TextWrapping="Wrap"
         />
```
```
    private void TextBox_PointerWheelChanged(object? sender, PointerWheelEventArgs e)
    {
        if (sender is TextBox textBox)
        {

            var scrollViewer = textBox.GetVisualDescendants()
                .OfType<ScrollViewer>()
                .FirstOrDefault();

            if (scrollViewer != null)
            {
                scrollViewer.Offset = new Avalonia.Vector(
                    scrollViewer.Offset.X, 
                    scrollViewer.Offset.Y - e.Delta.Y * 10); 
            }
        
            e.Handled = true; 
        }
    }
```

**Comment by tobyfirth** [comment]
You would need to target the padding of the internal `ScrollViewer` or `TextPresenter `(see [TextBox.xaml](https://github.com/AvaloniaUI/Avalonia/blob/d5f0188ccf9e36d9a3a6207bf4054deadf064f60/src/Avalonia.Themes.Fluent/Controls/TextBox.xaml#L139)).

Maybe something like this

```xml
<Style Selector="TextBox /template/ ScrollViewer#PART_ScrollViewer">
    <Setter Property="Padding" Value="0 0 32 0"/>
</Style>
```

You probably want to add a class selector or set the style just in the specific `TextBox`.

**Reply by TangDave** [reply]
Unfortunately I couldn't get this to work - it didn't make any difference.

**Reply by Wasdalt** [reply]
[Запись экрана от 2024-09-13 19-18-39.webm](https://github.com/user-attachments/assets/1b20a017-7a46-4bf0-bec2-e8ab3dfc6f2c)

**Reply by TangDave** [reply]
I ended up just setting 'AllowAutoHide' to false - this was the only way to not have the vertical scrollbar interfere with the text.

Your last example doesn't have a vertical scrollbar - I want the vertical scrollbar. The problem was that an auto-hiding vertical scrollbar in Avalonia hides part of the text.

I'll put a separate bug report in for this issue.

**Reply by Wasdalt** [reply]
![изображение](https://github.com/user-attachments/assets/21c99ba8-8910-4e9a-a01b-cdd02c0fe9f8)

**Reply by TangDave** [reply]
Yes - this is what I get when I set AllowAutoHide to false. But when AllowAutoHide is true, I don't see what you show.

**Reply by Wasdalt** [reply]
ScrollViewer.VerticalScrollBarVisibility="Visible"
